contracts/value/mod.go

```diff
@@ -0,0 +1,245 @@
+}
+
+// RegisterContract registers the value contract to the given execution service.
+func RegisterContract(exec *native.Service, c Contract) {
+	exec.Set(ContractName, c)
+}
+
+// Contract is a simple smart contract that allows one to handle the storage by
+// performing CRUD operations.
+//
+// - implements native.Contract
+type Contract struct {
+	// index contains all the keys set (and not delete) by this contract so far
+	index map[string]struct{}
+
+	// access is the access service to handle access on this smart contract
+	access access.Service
+
+	// accessKey is the identity's key allowed to use this smart contract
+	accessKey []byte
+
+	// cmd provides the commands executions
+	cmd commands
+
+	// printer is the output used by the READ and DISPLAY commands
+	printer io.Writer
+}
+
+// NewContract creates a new Value contract
+func NewContract(aKey []byte, srvc access.Service) Contract {
+	contract := Contract{
+		index:     map[string]struct{}{},
+		access:    srvc,
+		accessKey: aKey,
+		printer:   infoLog{},
+	}
+
+	contract.cmd = valueCommand{Contract: &contract}
+
+	return contract
+}
+
+// Execute implements native.Contract. It runs the appropriate command.
+func (c Contract) Execute(snap store.Snapshot, step execution.Step) error {
+	creds := NewCreds(c.accessKey)
+
+	err := c.access.Match(snap, creds, step.Current.GetIdentity())
+	if err != nil {
+		return xerrors.Errorf("identity not authorized: %v (%v)",
+			step.Current.GetIdentity(), err)
+	}
+
+	cmd := step.Current.GetArg(CmdArg)
+	if len(cmd) == 0 {
+		return xerrors.Errorf("'%s' not found in tx arg", CmdArg)
+	}
+
+	switch Command(cmd) {
+	case CmdSet:
+		err := c.cmd.set(snap, step)
+		if err != nil {
+			return xerrors.Errorf("failed to SET: %v", err)
+		}
+	case CmdRead:
+		err := c.cmd.read(snap, step)
+		if err != nil {
+			return xerrors.Errorf("failed to READ: %v", err)
+		}
+	case CmdDelete:
+		err := c.cmd.delete(snap, step)
+		if err != nil {
+			return xerrors.Errorf("failed to DELETE: %v", err)
+		}
+	case CmdDisplay:
+		err := c.cmd.display(snap)
+		if err != nil {
+			return xerrors.Errorf("failed to DISPLAY: %v", err)
+		}
+	default:
+		return xerrors.Errorf("unknown command: %s", cmd)
+	}
+
+	return nil
+}
+
+// valueCommand implements the commands of the value contract
+//
+// - implements commands
+type valueCommand struct {
+	*Contract
+}
+
+// set implements commands. It performs the SET command
+func (c valueCommand) set(snap store.Snapshot, step execution.Step) error {
+	key := step.Current.GetArg(KeyArg)
+	if len(key) == 0 {
+		return xerrors.Errorf("'%s' not found in tx arg", KeyArg)
+	}
+
+	value := step.Current.GetArg(ValueArg)
+	if len(value) == 0 {
+		return xerrors.Errorf("'%s' not found in tx arg", ValueArg)
+	}
+
+	err := snap.Set(key, value)
+	if err != nil {
+		return xerrors.Errorf("failed to set value: %v", err)
+	}
+
+	c.index[string(key)] = struct{}{}
+
+	dela.Logger.Info().Str("contract", "value").Msgf("setting %s=%s", key, value)
```

Comment:
If I understand their purpose correctly, this looks like `"contract"` and `"value"` could be defined as constants somewhere. Shouldn't we use `ContractName` ?